keps/sig-storage/20200709-pod-pvc-qos-limits.md

```diff
@@ -0,0 +1,255 @@
+- For volume capability of iops, it is a physical limitation on device(PV), and this is not the same as the QoS of PVC in this proposal.
+- As cgroup implement has its limitations, we will not mention kernel buffered writings issues with cgroup limitations. This should be fixed or optimized in kernel side. Detailes will be mentioned in the `risks and limitations` below.
+
+
+## Proposal
+
+After open a feature gate "VolumeQos=True" or "DeviceIOPS=true", pod volume qos can be added by annotation like below. The alpha MVP would use annotations.
+
+
+```
+annotations:
+    qos.volume.storage.daocloud.io: >-
+        {"pvc": "snap-03", "iops": {"read": 2000, "write": 1000}, "bps":
+        {"read": 1000000, "write": 1000000}}
+```
+
+
+Then kubelet can get the mount point of the pvc and the device id. Then we use the cgroup to limit iops and bps of the pod.
+We can just edit the cgroup limit files under the pod /sys/fs/cgroup/blkio/kubepods/pod/<Container_ID>/...
+For instance, to limit read iops of a pod
+
+
+```
+ echo "<block_device_maj:min> <value>" > /sys/fs/cgroup/blkio/kubepods/pod<UID>/blkio.throttle.read_iops_device
+```
+
+To manage the QoS of containers, it can be supported in the future if there are more than 1 contaienrs in the pod, and we may add different iops limits for each container.
+
+### Advantage of Proposal:
+
+* The implementation is based on pod dir, so this can be adapted to any Container Runtimes. If the implementation is based on dockerd iops limitation feature, other container runtime may would support this only if they implement similar features. 
+* This is a k8s way, not a docker way. From cluster aspect, application owner set the pv for pods and kubelet could detect which is mount point for the pod, but which device at runtime is mounted to the pod is not visable for application owner .
+
+This is the first/alpha MVP. If this is a suggested way to implement QoS of pvc, we can add iops/bps to container limits like cpu\memory.
+
+![image](https://user-images.githubusercontent.com/2010320/88526364-72201680-d02e-11ea-8dde-891c4f96ffaf.png)
+The difference is in step 5 that kubelet will read pod annotation and add iops limits by cgroup to pod
+
+### Beta implementation of this proposal:
+
+Add the limitation directly to volumes properties like below:
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: example
+  labels:
+    dce.daocloud.io/component: example
+spec:
+  containers:
+  - image: daocloud.io/daocloud/dao-2048:latest
+    name: local-2048-1
+    volumeMounts:
+    - mountPath: /data
+      name: volume
+    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
+      name: default-token-cv7mt
+      readOnly: true
+  serviceAccount: default
+  serviceAccountName: default
+  volumes:
+  - name: volume
+    persistentVolumeClaim:
+      claimName: local-storage-pvc-w-1
+    iops:
+      read: 2000
+      write: 1000
+    bps:
+      read: 1000000
+      write: 1000000
```

Comment:
1. This assumes that cluster admin can trust users to create pods with the right IOPS requirements. A rogue user can claim all / most of the IOPS for itself. In Kubernetes, we don't trust users (much) and we have quotas and similar concepts to limit them. I don't know how to apply IOPS there, perhaps via StorageClass & IOPS defined there? Then the cluster admin can define quota of volumes to each user / namespace.

2. How does it work when user sets different IOPS for volumes based by the same device (e.g. each Secret / EmptyDir / HostPath volume having different IOPS)?

3. Not all volume types support IOPS, e.g. NFS / GlusterFS / CephFS / tmpfs. Will be IOPS setting silently ignored for them?

Reply:
1. To limit user, we may add it to limitrange of namespace.
IOPS can be applied to storage class as well, but this may be a limitation from storage side. This may be a physical limitation. 

2. IOPS limitation may only apply to volume, which is a block device. 
cgroups can be applied to only device id.
If configmap and secrets are mounted on which is on the rootfs of the host, the rootfs iops limitation will take effect.

3. That's the limitation. 
Just ignore those in kubelet. kubelet can get mount type.

Reply:
Would it make sense to instead of adding absolute IOPS limits, add more of a QoS class to the storageclass. eg,

* unlimited, no restriction on IOPS, which is what we currently do

* fair, all pods using the PVC/volume split IOPS fairly. This does require a pretty complicated controller to track PV usage and update all the nodes with how usage should split, especially if we want to base throttling on actual usage (eg, pods that aren't hammering the device don't get the same share as a more active pod), but this may be what users actually want

* priority, pods with some sort of annotation get unlimited IOPS, all other pods get a fixed share

OK, I'll admit this is all very vague and not well thought-out, but a direction like this has a few advantages:

* Current state of the art means we're not going to be able to hit all IOPS limits anyway AFAIU

* This gives leeway for implementation differences and means we wouldn't have to change the API when cgroups v2 comes out or do all sort of complicated docker vs containerd differences

* Managing IOPS at a pod level is not going to scale for large deployments, but it's (probably) only in large deployments where this sort of thing is really needed. So a higher-level API is maybe ultimately much more useful.

#27000 already was discussing along these lines.

Maybe another way to see this is that QoS is really a volume characteristic and not a pod characteristic. That is, we are concerned about guaranteeing a certain performance for all users of a volume more than we are concerned about setting what performance a pod sees. So setting limits at some higher level than the pod is probably the right way to go about this.